# **Practical Questions:**

11. Imagine an accounting routine used in a book shop. It works on a list with sublists, which look like this:
Order Number	Book Title and Author	          Quantity	Price Per Item
34587	        Learning Python, Mark Lutz	        4	        40.95
98762	        Programming Python, Mark Lutz	      5	        56.80
77226	        Head First Python, Paul Barry	      3	        32.95
88112	        Einführung in Python3, Bernd Klein	3	        24.99

Write a Python program, which returns a list with 2-tuples. Each tuple consists of the order number and the product of the price per item and the quantity. The product should be increased by 10,- € if the value of the order is smaller than 100,00 €.

Write a Python program using lambda and map

Ans: You can create a Python program that processes the given data and returns a list of tuples, where each tuple contains the order number and the adjusted total price. The program will utilize lambda and map as requested. Here's how to implement this:




In [26]:
orders = [
    [34587, "Learning Python, Mark Lutz", 4, 40.95],
    [98762, "Programming Python, Mark Lutz", 5, 56.80],
    [77226, "Head First Python, Paul Barry", 3, 32.95],
    [88112, "Einführung in Python3, Bernd Klein", 3, 24.99]
]


Program to Calculate Adjusted Prices

In [ ]:
def calculate_adjusted_totals(orders):
    # Function to calculate total price and apply adjustment
    def calculate_order(order):
        order_number = order[0]
        quantity = order[2]
        price_per_item = order[3]
        total_price = quantity * price_per_item

        # Adjust the total price if it's less than 100
        if total_price < 100:
            total_price += 10

        return (order_number, total_price)

    # Use map to apply the calculation function to each order
    result = list(map(calculate_order, orders))
    return result

# Example usage
adjusted_totals = calculate_adjusted_totals(orders)
print(adjusted_totals)

Explanation:

Data Structure: The orders list contains sublists for each order with the specified fields.

calculate_order Function: This inner function computes the total price for each order and applies the €10 adjustment if the total is less than €100.

map: The map() function applies calculate_order to each sublist in orders.

Result: The result is converted to a list of tuples containing the order number and the adjusted total price.

Output
When you run the example usage, you would get output similar to:

In [ ]:
[(34587, 162.8), (98762, 284.0), (77226, 98.85), (88112, 69.97)]

10.Create a Python program that uses `filter()` to remove all the vowels from a given string.

Ans: You can use the filter() function in Python to remove all the vowels from a given string like this:


In [25]:
def remove_vowels(input_string):
    vowels = "aeiouAEIOU"
    # Use filter to keep only non-vowel characters
    filtered_string = ''.join(filter(lambda char: char not in vowels, input_string))
    return filtered_string

# Example usage:
input_string = "Hello, World!"
result = remove_vowels(input_string)
print(result)  # Output: Hll, Wrld!


Hll, Wrld!


9. Write a Python program that uses `map()` to convert a list of temperatures from Celsius to Fahrenheit.

Ans: You can use the map() function in Python to convert a list of temperatures from Celsius to Fahrenheit like this:

In [24]:
# Function to convert Celsius to Fahrenheit
def celsius_to_fahrenheit(celsius):
    return (celsius * 9/5) + 32

# List of temperatures in Celsius
celsius_temps = [0, 20, 37, 100]

# Using map to convert the list of temperatures
fahrenheit_temps = list(map(celsius_to_fahrenheit, celsius_temps))

# Example usage
print(fahrenheit_temps)


[32.0, 68.0, 98.6, 212.0]


8.Use a lambda function in Python to sort a list of tuples based on the second element of each tuple.

Ans: You can use a lambda function in Python to sort a list of tuples based on the second element of each tuple like this:

In [23]:
# Sample list of tuples
data = [(1, 'banana'), (2, 'apple'), (3, 'cherry'), (4, 'date')]

# Sorting the list based on the second element of each tuple
sorted_data = sorted(data, key=lambda x: x[1])

# Example usage
print(sorted_data)


[(2, 'apple'), (1, 'banana'), (3, 'cherry'), (4, 'date')]


7. Implement a generator function that reads a file line by line and yields each line as a string.

Ans: You can implement a generator function that reads a file line by line and yields each line as a string like this:

In [ ]:
def read_file_lines(file_path):
    with open(file_path, 'r') as file:
        for line in file:
            yield line.strip()  # Use strip() to remove leading/trailing whitespace

# Example usage:
file_path = 'example.txt'  # Replace with your file path
for line in read_file_lines(file_path):
    print(line)


Explanation:

Generator Function: The read_file_lines function opens the specified file and iterates through each line, yielding it after stripping any leading or trailing whitespace.

Example Usage: You can specify the path to your file, and the loop will print each line as it is read.

6.Write a generator function in Python that yields the powers of 2 up to a given exponent

Ans: You can create a generator function in Python that yields the powers of 2 up to a specified exponent like this:

In [13]:
def powers_of_two(exponent):
    for i in range(exponent + 1):
        yield 2 ** i

# Example usage:
exponent = 5
for power in powers_of_two(exponent):
    print(power)


1
2
4
8
16
32


Explanation:

Generator Function: The powers_of_two function uses the yield keyword to produce powers of 2 from
2
0
2
0
  to
2
exponent
2
exponent.

Example Usage: You can specify the desired exponent, and the loop will print each power of 2 up to that exponent. For an exponent of 5, it will yield 1, 2, 4, 8, 16, 32.

5. Create an iterator class in Python that generates the Fibonacci sequence up to a specified number of terms.

Ans: You can create an iterator class in Python that generates the Fibonacci sequence up to a specified number of terms like this:

In [12]:
class FibonacciIterator:
    def __init__(self, terms):
        self.terms = terms
        self.current = 0
        self.next_value = 1
        self.count = 0

    def __iter__(self):
        return self

    def __next__(self):
        if self.count < self.terms:
            self.count += 1
            self.current, self.next_value = self.next_value, self.current + self.next_value
            return self.current
        else:
            raise StopIteration

# Example usage:
fibonacci_terms = 10
fibonacci_sequence = FibonacciIterator(fibonacci_terms)

for number in fibonacci_sequence:
    print(number)


1
1
2
3
5
8
13
21
34
55


Explanation:
__init__ Method: Initializes the iterator with the desired number of terms and sets the starting values for the Fibonacci sequence.
__iter__ Method: Returns the iterator object itself.
__next__ Method: Calculates the next Fibonacci number. If the number of terms has been reached, it raises a StopIteration exception.

4. Write a Python function that checks if a given number is prime or not from 1 to 200.

Ans: You can create a Python function to check if a given number is prime, specifically for numbers from 1 to 200, like this:

In [11]:
def is_prime(n):
    if n <= 1:
        return False
    for i in range(2, int(n**0.5) + 1):
        if n % i == 0:
            return False
    return True

# Example usage:
for number in range(1, 201):
    if is_prime(number):
        print(f"{number} is prime.")


2 is prime.
3 is prime.
5 is prime.
7 is prime.
11 is prime.
13 is prime.
17 is prime.
19 is prime.
23 is prime.
29 is prime.
31 is prime.
37 is prime.
41 is prime.
43 is prime.
47 is prime.
53 is prime.
59 is prime.
61 is prime.
67 is prime.
71 is prime.
73 is prime.
79 is prime.
83 is prime.
89 is prime.
97 is prime.
101 is prime.
103 is prime.
107 is prime.
109 is prime.
113 is prime.
127 is prime.
131 is prime.
137 is prime.
139 is prime.
149 is prime.
151 is prime.
157 is prime.
163 is prime.
167 is prime.
173 is prime.
179 is prime.
181 is prime.
191 is prime.
193 is prime.
197 is prime.
199 is prime.


In this function, is_prime(n) checks if the number n is prime. It first handles the cases where n is less than or equal to 1. Then it checks for factors from 2 up to the square root of n. If any factor is found, the function returns False; otherwise, it returns True. The example usage prints out all prime numbers from 1 to 200.

3.Implement a Python function that takes a list of integers and returns a new list containing the squares of  each number.

Ans: You can implement a Python function to return a new list containing the squares of each integer in a given list like this:

In [10]:
def square_numbers(numbers):
    return [num ** 2 for num in numbers]

# Example usage:
input_list = [1, 2, 3, 4, 5]
result = square_numbers(input_list)
print(result)  # Output: [1, 4, 9, 16, 25]


[1, 4, 9, 16, 25]


2.Create a Python function that accepts a string and returns the reverse of that string.

Ans: You can create a Python function to reverse a string like this:

In [9]:
def reverse_string(s):
    return s[::-1]

# Example usage:
input_string = "Hello, World!"
result = reverse_string(input_string)
print(result)  # Output: !dlroW ,olleH

!dlroW ,olleH


1.Write a Python function that takes a list of numbers as input and returns the sum of all even numbers in the list.

Ans: def sum_even_numbers(numbers):
    even_sum = 0
    for num in numbers:
        if num % 2 == 0:
            even_sum += num
    return even_sum

In [8]:
def sum_even_numbers(numbers):
    return sum(num for num in numbers if num % 2 == 0)

# Example usage:
numbers = [1, 2, 3, 4, 5, 6]
result = sum_even_numbers(numbers)
print(result)  # Output: 12

12


# **Theory Questions:**

11.Using pen & Paper write the internal mechanism for sum operation using  reduce function on this given list:[47,11,42,13];

Ans: Here's the internal mechanism for the sum operation using the reduce function on the given list:

List: [47, 11, 42, 13]

Reduce Function: functools.reduce(lambda x, y: x + y, [47, 11, 42, 13])

Internal Mechanism:

1. Initialize x with the first element of the list: x = 47
2. Initialize y with the second element of the list: y = 11
3. Apply the lambda function: x = x + y = 47 + 11 = 58
4. Update x with the result: x = 58
5. Move to the next element in the list: y = 42
6. Apply the lambda function: x = x + y = 58 + 42 = 100
7. Update x with the result: x = 100
8. Move to the next element in the list: y = 13
9. Apply the lambda function: x = x + y = 100 + 13 = 113
10. Return the final result: 113

Step-by-Step Calculation:


47 + 11 = 58
58 + 42 = 100
100 + 13 = 113


Result: 113

10.What is the difference between `map()`, `reduce()`, and `filter()` functions in Python?

Ans: map(), reduce(), and filter() are three fundamental functions in Python's functional programming toolkit:

Map()

1. Applies a function to each element of an iterable.
2. Returns a map object (iterator).
3. Purpose: Transformation, data conversion.

Example:

numbers = [1, 2, 3, 4, 5]
squared_numbers = list(map(lambda x: x ** 2, numbers))
print(squared_numbers)  # [1, 4, 9, 16, 25]


Filter()

1. Applies a predicate function to each element of an iterable.
2. Returns a filter object (iterator) with elements for which the predicate is True.
3. Purpose: Data filtering, selection.

Example:

numbers = [1, 2, 3, 4, 5]
even_numbers = list(filter(lambda x: x % 2 == 0, numbers))
print(even_numbers)  # [2, 4]


Reduce()

1. Applies a binary function to all items in an iterable, going from left to right.
2. Returns a single output value.
3. Purpose: Accumulation, aggregation.

Example (using functools.reduce):

import functools
numbers = [1, 2, 3, 4, 5]
sum = functools.reduce(lambda x, y: x + y, numbers)
print(sum)  # 15
Key differences:

|  | Map() | Filter() | Reduce() |
| --- | --- | --- | --- |
| Purpose | Transformation | Filtering | Accumulation |
| Output | Iterator | Iterator | Single value |
| Function | Unary | Predicate | Binary |
| Operation | Apply | Select | Accumulate |


9. Explain the purpose and usage of the `map()` function in Python.

Ans: The map() function in Python:

Purpose:

The map() function applies a given function to each item of an iterable (such as a list, tuple, or string) and returns a map object, which is an iterator.

Syntax:


map(function, iterable)


Usage:

1. Transforming data: Apply a function to each element of an iterable.
2. Converting data types: Convert elements from one type to another.
3. Filtering data: Use a function to filter out unwanted elements.

Examples:

1. Square numbers:


numbers = [1, 2, 3, 4, 5]
squared_numbers = list(map(lambda x: x ** 2, numbers))
print(squared_numbers)  # [1, 4, 9, 16, 25]


1. Convert strings to uppercase:


strings = ["hello", "world", "python"]
uppercase_strings = list(map(str.upper, strings))
print(uppercase_strings)  # ["HELLO", "WORLD", "PYTHON"]


1. Filter out even numbers:


numbers = [1, 2, 3, 4, 5]
odd_numbers = list(map(lambda x: x if x % 2 != 0 else None, numbers))
odd_numbers = [x for x in odd_numbers if x is not None]
print(odd_numbers)  # [1, 3, 5]


Multiple Iterables:

map() can take multiple iterables as arguments. The function will be applied to the corresponding elements of each iterable.


numbers1 = [1, 2, 3]
numbers2 = [4, 5, 6]
sums = list(map(lambda x, y: x + y, numbers1, numbers2))
print(sums)  # [5, 7, 9]

8.What is a lambda function in Python and when is it typically used?

Ans: Lambda Function in Python:

A lambda function is a small, anonymous function that can take any number of arguments, but can only have one expression.

Syntax:


lambda arguments: expression


Example:


sum = lambda x, y: x + y
print(sum(3, 4))  # Output: 7


When to Use Lambda Functions:

1. One-time use: Lambda functions are ideal for one-time use cases, eliminating the need for a named function.
2. Simple operations: Use lambda for simple operations, like data transformations or filtering.
3. Event handling: Lambda functions are useful as event handlers, such as button clicks or keyboard events.
4. Map(), Filter(), Reduce(): Lambda functions are often used with these functions for data processing.
5. Sorting: Use lambda as the key function for custom sorting.
6. Data analysis: Lambda functions are helpful in data analysis, especially with libraries like Pandas.

Benefits:

1. Concise code: Lambda functions reduce code length.
2. Readability: They improve readability by encapsulating simple logic.
3. Flexibility: Lambda functions can be defined anywhere.

Common Use Cases:

1. Sorting a list of dictionaries by a specific key:

students = [{'name': 'John', 'age': 20}, {'name': 'Alice', 'age': 22}]
students.sort(key=lambda x: x['age'])

1. Filtering a list:

numbers = [1, 2, 3, 4, 5]
even_numbers = list(filter(lambda x: x % 2 == 0, numbers))

1. Mapping a list:

numbers = [1, 2, 3, 4, 5]
squared_numbers = list(map(lambda x: x ** 2, numbers))

7.What are the advantages of using generators over regular functions?

Ans: Advantages of Generators over Regular Functions:

Memory Efficiency

1. Generators store only the current state, reducing memory usage.
2. Ideal for handling large datasets or infinite sequences.

Improved Performance

1. Generators avoid computing unnecessary values.
2. Faster iteration due to lazy evaluation.

Flexibility

1. Generators can produce values on-the-fly.
2. Suitable for real-time data processing or cooperative multitasking.

Efficient Iteration

1. Generators eliminate the need for indexing or slicing.
2. Natural fit for for loops or other iterative constructs.

Better Error Handling

1. Generators can handle errors more elegantly.
2. Easy to implement retry mechanisms or error handling.

Enhanced Code Readability

1. Generators encapsulate complex logic.
2. Simplify code by breaking down complex operations.

Comparison with Regular Functions

|  | Regular Functions | Generators |
| --- | --- | --- |
| Memory Usage | Stores all values | Stores only current state |
| Computation | Computes all values | Computes values on-the-fly |
| Iteration | Returns all values | Yields values one-by-one |
| Error Handling | More complex | Easier to handle |
| Code Readability | More verbose | More concise |

Use Generators When

1. Handling large datasets.
2. Processing real-time data.
3. Implementing cooperative multitasking.
4. Improving performance.
5. Simplifying complex logic.

Example

# Regular function
def squares(n):
    result = []
    for i in range(n):
        result.append(i ** 2)
    return result

# Generator function
def squares_gen(n):
    for i in range(n):
        yield i ** 2

# Usage
for square in squares_gen(10):
    print(square)

6.Explain the concept of generators in Python and how they are defined.

Ans: Generators in Python:

What are Generators?

Generators are a type of iterable, like lists or tuples, but they don't store all values in memory at once. Instead, they generate values on-the-fly as needed.

Key Characteristics:

1. Lazy Evaluation: Values are computed only when needed.
2. Memory Efficiency: No need to store all values in memory.
3. Iteratable: Can be looped through using for loops.

Defining Generators:

Generators are defined using functions with the yield keyword.

Basic Syntax:

def generator_name():
    # Code to generate values
    yield value1
    # More code to generate values
    yield value2
    # ...

Example:

def infinite_sequence():
    n = 0
    while True:
        yield n
        n += 1

# Create generator object
gen = infinite_sequence()

# Iterate over generator
print(next(gen))  # 0
print(next(gen))  # 1
print(next(gen))  # 2

5. What are iterators in Python and how do they differ from iterables?

Ans: Iterators and iterables are fundamental concepts in Python:

Iterables

1. Objects that can be iterated over (e.g., lists, tuples, dictionaries, sets).
2. Can be looped through using for loops or converted to iterators.
3. Can be iterated over multiple times.

Iterators

1. Objects that keep track of their position during iteration.
2. Created from iterables using the iter() function.
3. Can only be iterated over once.

Example:

# Iterable (list)
my_list = [1, 2, 3]

# Create iterator from iterable
my_iter = iter(my_list)

# Iterate over iterator
print(next(my_iter))  # 1
print(next(my_iter))  # 2
print(next(my_iter))  # 3

# Try to iterate again (will raise StopIteration)
try:
    print(next(my_iter))
except StopIteration:
    print("Iterator exhausted")

# Iterate over iterable again (no issue)
for num in my_list:
    print(num)  # 1, 2, 3

4.What is the purpose of the `return` statement in a Python function?

Ans: The return statement in a Python function serves several purposes:

1. Exiting the Function: The return statement terminates the function's execution and returns control to the caller.

2. Returning Values: The return statement can pass values back to the caller. These values can be variables, literals, or expressions.

3. Default Return Value: If a function doesn't explicitly use the return statement, it will implicitly return None.

4. Multiple Return Values: Python allows functions to return multiple values using tuples, lists, or dictionaries.

5. Early Termination: The return statement can be used to exit a function prematurely, skipping remaining code.

Example:

def greet(name):
    print(f"Hello, {name}!")
    return "Greeting completed"

result = greet("John")
print(result)  # Output: Greeting completed

3.What are the different ways to define and call a function in Python?

Ans: Here are the different ways to define and call a function in Python:

Defining Functions

1. Function Definition: Using the def keyword

def greet(name):
    print(f"Hello, {name}!")

greet("John")

2. Lambda Functions: Anonymous functions

greet = lambda name: print(f"Hello, {name}!")
greet("John")

3. Nested Functions: Functions inside functions

def outer():
    def inner(name):
        print(f"Hello, {name}!")
    inner("John")
outer()


Calling Functions

1. Positional Arguments: Arguments passed in order

def greet(name, age):
    print(f"Hello, {name}! You're {age}.")
greet("John", 30)

2. Keyword Arguments: Arguments passed by keyword

greet(name="John", age=30)

3. Mixed Arguments: Combination of positional and keyword

greet("John", age=30)

4. Default Argument Values: Optional arguments with default values

def greet(name, age=30):
    print(f"Hello, {name}! You're {age}.")
greet("John")

5. Variable Arguments: Arbitrary number of arguments (*args, **kwargs)

def flexible_greet(*args, **kwargs):
    print("Arguments:", args)
    print("Keyword Arguments:", kwargs)
flexible_greet("John", 30, location="New York")

2.Explain the concept of function arguments and parameters in Python.

Ans: In Python, function arguments and parameters are essential concepts:

Parameters:

1. Defined in the function definition.
2. Receive values passed during function calls.
3. Can have default values.
4. Can be positional or keyword-based.

Arguments:

1. Values passed to a function during a call.
2. Can be literals, variables, or expressions.
3. Matched with parameters based on position or keyword.

Types of Parameters:

1. Positional Parameters: Defined without keywords.
2. Keyword Parameters: Defined with keywords.
3. Default Parameters: Have default values.
4. Variable Parameters: Accept arbitrary numbers of arguments (*args, **kwargs).

Example:

def greet(name, age=30, location="New York"):
    print(f"Hello, {name}! You're {age} from {location}.")

# Positional arguments
greet("John", 25)

# Keyword arguments
greet(name="Jane", age=35, location="London")

# Mixing positional and keyword arguments
greet("Bob", location="Paris", age=40)

Variable Parameters:

1. *args: Accepts arbitrary positional arguments.
2. **kwargs: Accepts arbitrary keyword arguments.

Example:

def flexible_greet(*args, **kwargs):
    print("Arguments:", args)
    print("Keyword Arguments:", kwargs)

flexible_greet("John", 25, location="New York", occupation="Developer")

1.What is the difference between a function and a method in Python?

Ans: In Python, functions and methods are both blocks of code that perform specific tasks, but they differ in their context and usage:

Functions:

1. Standalone blocks of code.
2. Defined outside classes.
3. Can be called directly.
4. Do not have access to class attributes.

Example:


def greet(name):
    print(f"Hello, {name}!")

greet("John")


Methods:

1. Blocks of code bound to a class or object.
2. Defined inside classes.
3. Called on instances of the class.
4. Have access to class attributes (data members).

Example:


class Person:
    def __init__(self, name):
        self.name = name

    def greet(self):
        print(f"Hello, my name is {self.name}!")

john = Person("John")
john.greet()